### ISSUE TITLE
braille and no-break-space

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#1717]
**Reported by andre14 on 2011-08-11 22:30**
NVDA does not transcribed breaking spaces when the cursor is not on the sign.
In other words, if the cursor is elsewhere than on the non-breaking space, NVDA transcribe it by a normal space.

**Comment by DrSooom** [comment]
### Actual behaviour in NVDA 2018.1:
1. This issue still exists in NVDA 2018.1 with "de-de-comp8.ctb" on win7x64. this table contains "sign \x00a0 7", but U+00A0 (= no-break space) is only shown on the braille display when "Expand to computer braille for the word at the cursor" is enabled and when the cursor is localed on the left side of this character.
2. The braille table "fr-bfu-comp8.utb" shows dot 7 for no-break spaces all the time, whehter "Expand to computer braille for the word at the cursor" is enabled or disabled. And it also doesn't matter where the cursor is located.
3. And "UEBC-g1.utb" is the complete opposite of "fr-bfu-comp8.utb". "space \X00A0 0" can be found in "ukchardefs.cti". Changing this value in "ukchardefs.cti" to "space \X00A0 7" has no effect – also after restarting NVDA, which is a little bit strange.
### Expected behaviour for German 8-dot table:
No-break spaces in "de-de-comp8.ctb" should be shown all the time, wherever the cursor is located, with dot 7.
#### Note:
You can enter a no-break space in Microsoft Word 2010 with Shift+CTRL+Space oand in HTML with "& nbsp ;" (without the spaces).
### Next steps:
@jcsteh or @leonardder: Shall I upload a logfile for this? And please re-open and rename this issue because it isn't fixed yet (quick test for "de-de-comp8.ctb" in NVDA 2018.3 Beta 3). Thanks.

**Comment by Andre9642** [comment]
Just add this line in de-de-comp8.ctb:

> always \x00A0 7

Result:

```
$ echo "a b" | lou_trace de-de-comp8.ctb
⠁⡀⠃
1.      sign    a       1
2.      always          7
3.      sign    b       12
```

I can create a PR for liblouis if needed.

**Comment by DrSooom** [comment]
Replacing `sign \x00a0 7` with `always \x00a0 7` in "de-de-comp8.ctb" solved this issue. Thanks a lot for this hint.

And yes, please create a PR for it because on iOS no-break spaces are always shown as ⡀ (U+2840, dot 7). It should be the same in NVDA (and other applications which are using liblouis).

**Comment by DrSooom** [comment]
Important update: Do not replace `sign \x00a0 7` with `always \x00a0 7` because if "Expand to computer braille for the word at the cursor" is enabled and the cursor located on the left side of a no-break space, the dots 7 and 8 are blinking together. This would be a new issue. So just add, as you already written, `always \x00a0 7` directly below `sign \x00a0 7`. Then the behaviour will be the same as before – only dot 8 is blinking.

**Comment by DrSooom** [comment]
Adding `sign \x2003 7`, `sign \x2006 7` and `sign \x2009 7` would be nice too. Currently I have changed/added [50 items in this German table](https://github.com/nvaccess/nvda/files/2342963/Braille-Tabelle_DE-8-Punkte.txt). So adding them could be helpful for all the others who not editing their braille tables.

**Comment by Andre9642** [comment]
> Adding sign \x2003 7, sign \x2006 7 and sign \x2009 7 would be nice too.

U+2003, U+2006 and U+2009 are not non-breaking space variants. However, I suggest U+202F (narrow no-break space).

**Comment by DrSooom** [comment]
You are right. Those three characters are other types of spaces.

A&#x202f;B (`A&#x202f;B`, U+202F) must be added to all "symbols.dic" in NVDA (or just for the German one, needs more deeper testing). I added this character right now in my NVDA setup. Thanks for the hint.

**Comment by DrSooom** [comment]
U+F0B7 is a bullet list character used in Microsoft Office. Please don't ask me why.

I only know the PDF file called [Das System der deutschen Blindenschrift](https://www.pharmabraille.com/wp-content/uploads/2015/01/system_d_blindenschrift_7620.pdf). I only took a look at the first 40 pages (or so) in the end of 2012. The rest didn't – and still doesn't – interest me.

Furthermore since liblouis 3.0.0 (or 3.1.0) many characters in the "de-de-comp8.ctb" were gone in compare to the 2.x versions. Based on [this commit](https://github.com/liblouis/liblouis/commit/cc72133b6dcf2a6d72a5657d7a8793f1e343db49#diff-eedf261b6f49c4a8376a16af60a41171) there could be a licencing problem with those characters. So my table contains just a few of those characters again, because I really need them – and they are also quite often using in German like …, ’, ” and €. To be honest, deleting them from the table was an annoying, huge mistake. Getting them back (maybe I'm doing this by my own) would be great, otherwise everyone (advanced user) will use their own definitions.

In other words: I don't know a German standard for the 8 dot table. I modified this table that it fits my needs best. Personally I would prefer to add custom characters to an additional user-based braille table via the NVDA GUI similar to the "symbols.dic". See issue #291, #3304 and #4996 and PR #5489.

So, if you want to be on the secure side, just add the spaces characters mentioned above. And please label this issue with "needsExternalFix" and rename it from "braille and no-break-space" to "de-de-comp8.ctb: No-break spaces (U+00A0) aren't shown all the time" (or something like that) – as the original author seems not to be reachable anymore. This issue will be closed by a NVDA-liblouis-update-PR later. Thanks.

**Comment by Adriani90** [comment]
@Andre9642 I see that the pull request referenced above is merged to master. In which version of Liblouis will this be solved? Is this already Liblouis 3.9 which has now been updated in NVDA?

**Comment by Andre9642** [comment]
@Adriani90 yes, the original issue is now solved in libLouis. You can close this.

**Comment by Adriani90** [comment]
closing as works for me.
